### PR TITLE
Check for broken links in markdown files

### LINKED ISSUE #312: Markdown Link Verification
Since v1.0 a lot of files have moved around to make place for new functionality for example extra example packs or simply neatening up the file structure and this causes links to break.

While it seems the structure is a bit more stable now and links should continue to work I was wondering if we should consider some sort of check in travis?

By no means am I saying use this, but [this CLI tool](https://www.npmjs.com/package/markdown-link-check) could work by just checking each of the our MD files causing builds to fail if they produce broken links catching them before they get merged!

### PR DESCRIPTION
**Pull request by Sherlouk**
## Changes in this pull request

- Install `markdown-link-check` as a dependency
- Check each markdown file for links, and check them using aforementioned library

Closes #312 

A few points of note, I'm not sure if we can use the built-in caching capabilities to cache the library (at the moment it's installed for each build, even though it's only used for 1 -- seems a big waste!). Also a downside of the library is it doesn't really like relative links, we only had one -- so I changed it to be absolute.

## Pull request checklist

- [x] All tests pass. Demo project builds and runs.
- [x] I added tests, an experiment, or detailed why my change isn't tested.
- [x] I have reviewed the [contributing guide](https://github.com/Instagram/IGListKit/blob/master/.github/CONTRIBUTING.md)

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -39,15 +39,21 @@ env:
 
 before_install:
   - gem install slather --no-rdoc --no-ri --no-document --quiet
+  - npm install -g markdown-link-check
 
 script:
 - set -o pipefail
 
+- if [ $CHECK_MARKDOWN == "YES" ]; then
+      echo -en 'travis_fold:start:script.1\\r';
```
Comment (Sherlouk) [inline .travis.yml]: It prints out a big block with ticks and crosses next to each URL, added a fold so it doesn't take over the entire job window.

File: CHANGELOG.md [hunk 2]
```diff
@@ -51,7 +51,7 @@ You can find a [migration guide here](https://instagram.github.io/IGListKit/migr
 
 - Added support for cells created from nibs. [Sven Bacia](https://github.com/svenbacia) [(#56)](https://github.com/Instagram/IGListKit/pull/56)
 
-- Added an additional initializer for `IGListSingleSectionController` to be able to support single sections created from nibs. An example can be found [here](Examples/Examples-iOS/IGListKitExamples/ViewControllers/SingleSectionViewController.swift). [(#56)](https://github.com/Instagram/IGListKit/pull/56)
+- Added an additional initializer for `IGListSingleSectionController` to be able to support single sections created from nibs. An example can be found [here](https://github.com/Instagram/IGListKit/tree/master/Examples/Examples-iOS/IGListKitExamples/ViewControllers/SingleSectionViewController.swift). [(#56)](https://github.com/Instagram/IGListKit/pull/56)
```
Comment (jessesquires) [inline CHANGELOG.md]: 👍

**Comment by jessesquires** [comment]
Thanks @Sherlouk !

I'd also like to see an example of a failure, if possible.

File: .travis.yml (hunk 1 above)
Comment (jessesquires) [inline .travis.yml]: actually, it's probably ok *not* to fold this. it's only going to run on the first build, right?

doesn't travis fold most steps automatically after completion?

**Comment by Sherlouk** [comment]
@jessesquires Will wait for this build to go through fine (moved the `npm install`). Assuming that works I'll break a link and let that build. (then probably forget to fix it 😂)

File: .travis.yml (hunk 1 above)
Comment (Sherlouk) [inline .travis.yml]: Travis does to some auto-folding, but from the testing I've done it never seemed to fold this.
Yes it only gets triggered on the first build, I'll remove it for the next build and can see if you're happy to keep it unfolded after that!

File: .travis.yml [hunk 3]
```diff
@@ -39,15 +39,21 @@ env:
 
 before_install:
   - gem install slather --no-rdoc --no-ri --no-document --quiet
+  - npm install -g markdown-link-check
```
Comment (Sherlouk) [inline .travis.yml]: Moved the `npm install`, and I've googled around a lot with regards to the `cache:` and you can add a bit which is `directories:` so I tried specifying `node_modules` but I couldn't see this effecting anything.

Happy to take input though!

**Comment by Sherlouk** [comment]
Glad we tested that, it seems the `set -o pipefail` was making it do the opposite of what we want it to do on this command. Rather than changing it for all the others, I've made it exit (not echo, as my testing instead had oops) immediately on fail.

Also for what it's worth the list of files this will check:
>./.github/CONTRIBUTING.md
./.github/ISSUE_TEMPLATE.md
./.github/PULL_REQUEST_TEMPLATE.md
./CHANGELOG.md
./Examples/Examples-iOS/LICENSE-examples.md
./Examples/Examples-tvOS/LICENSE-examples.md
./Guides/Getting Started.md
./Guides/IGListDiffable and Equality.md
./Guides/Migration.md
./LICENSE.md
./PATENTS.md
./Pods/OCMock/README.md
./README.md

**Comment by Sherlouk** [comment]
Hurrah! So after eliminating multiple variables it turns out the reason a bad link passed the built was because of the folding I added... Intriguing!

Alas it is working now!

**Comment by jessesquires** [comment]
Nice! 💯 Looks awesome. 🎉 Thanks @Sherlouk !

No need to squash, our syncing system will take care of that.

**Comment by jessesquires** [comment]
Oh - let's fix that bad link 😄

**Comment by rnystrom** [comment]
This is really awesome. TIL!

**Pull request closed without merging**